File: ruby/frankenstein_driver.rb
require  'socket'
class Regexp
  def to_s
    "regex:" + source
  end
end

# Convenient constants.
ANYTHING = /.*/

# The Frankenstein driver allows a Swing user interface to be tested.
# Regular expressions can be specified using the regular Ruby syntax.
module FrankensteinDriver
  @@test_dir =""
  attr_accessor :test_status

  # Activates a dialog with a specified title.
  #
  # The title can be specified as a regular expression
  def activate_dialog(title)
    append_to_script "activate_dialog \"#{title}\""
  end

  # Activates a window with a specified title.
  #
  # The title can be specified as a regular expression
  def activate_window(title)
    append_to_script "activate_window \"#{title}\""
  end

  # Activates an internal frame.
  #
  # The title can be specified as a regular expression
  def activate_internal_frame(title)
    append_to_script "activate_internal_frame \"#{title}\""
  end

  # Check whether a check box with the specified name is selected
  def assert_checkbox_selected(checkbox_name)
    assert_togglebutton_selected checkbox_name
  end

  # Check whether a check box with the specified name is not selected
  def assert_checkbox_not_selected(checkbox_name)
    assert_togglebutton_not_selected checkbox_name
  end

  # Check whether a radio button with the specified name is selected
  def assert_radiobutton_selected(radiobutton_name)
    assert_togglebutton_selected radiobutton_name
  end

  # Check whether a radio button with the specified name is not selected
  def assert_radiobutton_not_selected(radiobutton_name)
    assert_togglebutton_not_selected radiobutton_name
  end

  # Check whether a toggle button with the specified name is selected
  def assert_togglebutton_selected(togglebutton_name)
    assert_true togglebutton_name,"selected"
  end

  # Check whether a toggle button with the specified name is not selected
  def assert_togglebutton_not_selected(togglebutton_name)
    assert_false togglebutton_name,"selected"
  end

  # Check whether a component with the specified name is enabled
  def assert_enabled(component_name)
    assert_true component_name,"enabled"
  end

  # Check whether a component with the specified name is enabled
  def assert_disabled(component_name)
    assert_false component_name,"enabled"
  end

  # Check the number of rows displayed in a table
  def assert_number_of_table_rows(table_name, number_of_rows)
    assert table_name,"rowCount",number_of_rows
  end

  # Check the value of a specified table cell. The cell value can be specified as a regular expression
  def assert_table_cell(table_name, row, column, value)
    assert table_name, "getValueAt(#{row},#{column})", value
  end

  # Check the elements of a specified table row.
  #
  # The cell values can be an arbitrary list of arguments, which are checked from left to
  # right, starting at the first column.
  #
  # The cell values can be specified as regular expressions.
  def assert_table_row(table, row, *cell_values)
    cell_values.each_index {|index| assert_table_cell(table, row, index, cell_values[index])}
  end

  # Checks whether an OGNL expression evaluated against a specified component matches the specified value.
  #
  # The expected value can be specified as a regular expression.
  def assert(component_name,ognl_expression,expected_value)
    append_to_script "assert \"#{component_name}\" \"#{ognl_expression}:#{expected_value}\""
  end

  # Checks whether the specified OGNL expression is true.
  def assert_true(component_name,ognl_expression)
    assert component_name,ognl_expression,"true"
  end

  # Checks whether the specified OGNL expression is false.
  def assert_false(component_name,ognl_expression)
    assert component_name,ognl_expression,"false"
  end

  # Checks whether the text of a specified text component matches the specified value.
  #
  # The expected text can be specified as a regular expression.
  def assert_text(text_component, expected_text)
    assert text_component,"text",expected_text
  end

  # Cancels a table edit.
  def cancel_table_edit(table)
    append_to_script "cancel_table_edit \"#{table}\""
  end

  # Click on a specified button
  def click_button(button)
    append_to_script "click_button \"#{button}\""
  end

  # Click on a specified checkbox.
  #
  # The specified value can be true or false, and will select or deselect the checkbox
  def click_checkbox(button, value)
    append_to_script "click_checkbox \"#{button}\" \"#{value}\""
  end

  # Click on a specified radio button.
  def click_radiobutton(button)
    append_to_script "click_radio_button \"#{button}\""
  end

  # Click on a specified column of a table header
  def click_table_header(header_name,column_name)
    append_to_script "click_table_header \"#(header_name)\" \"#(column_name)\""
  end

  # Close an internal frame with the specified title.
  #
  # The title can be specified as a regular expression.
  def close_internal_frame(title)
    append_to_script "close_internal_frame \"#{title}\""
  end

  # Close all open dialogs recursively until no more dialogs are open.
  #
  # This function can be used to remove unexpected dialogs at the end of a test.
  def close_all_dialogs
    append_to_script "close_all_dialogs"
  end

  # Wait for a specified duration (in milliseconds)
  def delay(duration_in_milliseconds)
    append_to_script "delay \"#{duration_in_milliseconds}\""
  end

  # Wait for a dialog with a specified title to be closed.
  #
  # This function waits for 10 seconds in case the dialog is found to be open.
  # (The wait will be made configurable in a future release).
  # The title can be specified as a regular expression.
  def dialog_closed(title)
    append_to_script "dialog_closed \"#{title}\""
  end

  # Wait for a dialog with a specified title to be opened.
  #
  # This function waits for 10 seconds in case the dialog is not found to be open.
  # (The wait will be made configurable in a future release).
  # The title can be specified as a regular expression.
  def dialog_shown(title)
    append_to_script "dialog_shown \"#{title}\""
  end

  # Double click on a table row of a specified table.
  def double_click_table_row(table, row_index)
    append_to_script "double_click_table_row \"#{table}\",\"#{row_index}\""
  end

  # Double click on a list item of a specified list.
  def double_click_list(list, row_index)
    append_to_script "double_click_list \"#{list}\",\"#{row_index}\""
  end

  # Double clicks on a tree. Supports regular expressions
  #
  # For example: double_click_tree "tree_name", "top level", /.*level/, "third level"
  def double_click_tree(tree,*path)
    path = tree_path(path_elements)
    append_to_script "double_click_tree \"#{tree}\",#{path}"
  end

  # Enter the specified text into a specified text field.
  def enter_text(textfield, text)
    append_to_script "enter_text \"#{textfield}\",\"#{text}\""
  end

  # Edit a table cell at the specified coordinates. The coordinates are specified as a
  # "<row>,<column>" - for example, "1,1"
  def edit_table_cell(table, coords)
    append_to_script "edit_table_cell \"#{table}\",\"#{coords}\""
  end

  # Checks that an internal frame with the specified title has been shown
  # The title can be specified as a regular expression.
  def internal_frame_shown(title)
    append_to_script "internal_frame_shown \"#{title}\""
  end

  # Enter a keystroke with the specified modifiers.
  #
  # This method isn't too tester friendly at the moment (the modifiers and the keycode are both integers).
  # It is recommended that the recorder be used to create keystroke steps.
  # Future releases will make it easier to specify keystrokes.
  def key_stroke(modifiers, keycode)
    append_to_script "key_stroke \"#{modifiers},#{keycode}\""
  end

  # Navigate to a specified path in a menu item.
  # Both menu bars and popup menus are supported.
  #
  # The path needs to be specified as a string delimited by the > character.
  # For example: "first level>second level>item".
  # This function does not have regular expression support.
  def navigate(path)
    append_to_script "navigate \"#{path}\""
  end

  # Right click on a tree item of the specified tree. Supports regular expressions
  #
  # For example: right_click_tree "tree_name", "top level", /.*level/, "third level"
  def right_click_tree(tree, *path_elements)
    path = tree_path(path_elements)
    append_to_script "right_click_tree \"#{tree}\",#{path}"
  end

  # Right click on a list item of a specified list.
  def right_click_list(list,item_index)
    append_to_script "right_click_list \"#{list}\",\"#{item_index}\""
  end

  # Right click on a table row of a specified table.
  def right_click_table_row(table,row_index)
    append_to_script "right_click_table_row \"#{table}\",\"#{row_index}\""
  end

  # Select a specified value from a specified combo box.
  def select_drop_down(combo, value)
    append_to_script "select_drop_down \"#{combo}\",\"#{value}\""
  end

  # Select a specified file in a file chooser.
  def select_file(path)
    append_to_script "select_file \"#{path}\""
  end

  # Select multiple files in a file chooser.
  def select_files(paths)
    append_to_script "select_files \"#{paths}\""
  end

  # Select a specified value in a specified list.
  def select_list(list, value)
    append_to_script "select_list \"#{list}\",\"#{value}\""
  end

  # Selects the specified rows in a table.
  # rows is a comma separated list of table rows.
  #
  # Example: select_table_row "table_name" , "1,2,3"
  def select_table_row(table, rows)
    append_to_script "select_table_row \"#{table}\",\"#{rows}\""
  end

  # Select a specified tree path. Supports regular expressions
  #
  # For example: select_tree "tree_name", "top level", /.*level/, "third level"
  def select_tree(tree, *path_elements)
    path = tree_path(path_elements)
    append_to_script "select_tree \"#{tree}\",#{path}"
  end

  # Stop editing a specified table.
  def stop_table_edit(table)
    append_to_script "stop_table_edit \"#{table}\""
  end

  # Switch to a tab with the specified title of a specified tabbed pane.
  # The title can be specified as a regular expression.
  def switch_tab(tab, title)
    append_to_script "switch_tab \"#{tab}\",\"#{title}\""
  end

  # Moves a specified slider to the specified position
  def move_slider(slider,position)
    append_to_script "move_slider \"#{slider}\",\"#{position}\""
  end

  # Sends a test script to the Frankenstein Java runtime at the specified host and port.
  #
  # Waits for the test to complete, and reports test results.
  def run_frankenstein_test(host,port)
    init host,port
    start_test @@test_dir == "" ? @test_name : @@test_dir + "/" + @test_name
    test
    finish_test
    puts @test_name + " : " + (@test_status == "F" ? "Failed" : "Passed")
    @test_status
  end

  # Sets the location of the Frankenstein test reporting directory. The directory is usually specified before a test run.
  def FrankensteinDriver.test_dir=(dir)
    @@test_dir = dir
  end

  # Returns the location of the current test reporting directory.
  def FrankensteinDriver.test_dir
    @@test_dir
  end

  private
  def init(host,port)
    @test_name = self.class.to_s
    @host = host
    @port = port
    @script = ""
  end

  def append_to_script(script_line)
    @script += script_line.gsub("\n", "&#xA;") + "\n"
  end

  def start_test(testname)
    @script +="start_test \"#{testname}\"\n"
  end

  def finish_test
    socket = TCPSocket.new(@host,@port)
    socket.write @script
    socket.close_write
    recvthread = Thread.start do
        @test_status = socket.read
      end
    recvthread.join
    socket.close
  end

  def tree_path(*path_elements)
    "\"" + path_elements.join("\",\"") + "\""
  end
end

# Runs multiple tests in a suite. Generates an index file that displays whether a test passed or failed.
class TestRunner
  def initialize(host="localhost",port=5678)
    @test_reporter = TestReporter.new
    @host = host
    @port = port
  end

  # Runs multiple tests. The argument to the function is a list of Frankenstein test classes.
  def run(*tests)
    tests.each {|test| @test_reporter.report_test_result(test,test.new.run_frankenstein_test(@host,@port))}
    @test_reporter.report
  end
end

# Stores the result of a test run.
class TestResult
  attr_accessor :test
  def initialize(test,status)
    @test,@status = test,status
  end

  def status
    @status == "P" ? "#CFFFCF" : "#FFCFCF"
  end
end

# Creates an HTML report of a test run. Lists out the names of the tests that ran, and color codes tests that passed or failed.
class TestReporter
  def initialize
    @tests = []
  end

  def report_test_result(test,result)
    @tests.push TestResult.new(test,result)
  end

  def report
    index_file = File.new(FrankensteinDriver.test_dir + "/" + "index.html", "w+")
    index_file.puts "<html>"
    index_file.puts "<head><title>Test Results</title></head>"
    index_file.puts "<body>"
    index_file.puts "<h3>Test Results</h3><br>"
    index_file.puts "<table BORDER CELLSPACING=0 CELLPADDING=4>"
    @tests.each do |value|
      index_file.puts "<tr><td bgcolor=#{value.status}><font size=2 color=black><a href=\"#{value.test}.html\">#{value.test}</a></font></td></tr>"
    end
    index_file.puts "</table>"
    index_file.puts "</body>"
    index_file.puts "</html>"
    index_file.close
  end
end

File: src/com/thoughtworks/frankenstein/drivers/FrankensteinDriver.java
package com.thoughtworks.frankenstein.drivers;

/**
 * The Frankenstein driver allows a Swing user interface to be tested.
 * <p/>
 * <p/>
 * The Client of this API should ensure that finishTest() method is called after doing
 * all test assertions without fail.
 * </p>
 *
 * @author Pavan
 * @author Prakash
 */
public interface FrankensteinDriver {

    /**
     * Activates a dialog with a specified title.
     * The title can be specified as a regular expression.
     *
     * @param dialogTitle The title of the dialog to be activated.
     */
    public void activateDialog(String dialogTitle);

    /**
     * Activates a window with a specified title.
     * <p/>
     * The title can be specified as a regular expression.
     *
     * @param windowTitle The title of the window to be acitvated.
     */
    public void activateWindow(String windowTitle);

    /**
     * Activates an internal frame.
     * The title can be specified as a regular expression.
     *
     * @param internalFrameTitle The title of the internal frame to be activated.
     */
    public void activateInternalFrame(String internalFrameTitle);

    /**
     * Checks whether an OGNL expression evaluated against a specified component matches the specified value.
     * The expected value can be specified as a regular expression.
     *
     * @param componentName  The component name.
     * @param ognlExpression The OGNL expression to be evaluated.
     * @param expectedValue  The expected value.
     */
    public void assertValue(String componentName, String ognlExpression, String expectedValue);

    /**
     * Check the number of rows displayed in a table
     *
     * @param tableName            The name of the table.
     * @param expectedNumberOfRows The expected number of rows in the table.
     */
    public void assertNumberOfTableRows(String tableName, int expectedNumberOfRows);

    /**
     * Check the value of a specified table cell. The cell value can be specified as a regular expression.
     *
     * @param tableName         The name of the table.
     * @param row               The row number of the cell.
     * @param column            The column number of the cell.
     * @param expectedCellValue The expected cell value.
     */
    public void assertTableCell(String tableName, int row, int column, String expectedCellValue);

    /**
     * Check the elements of a specified table row.
     * <p/>
     * The cell values can be an arbitrary list of arguments, which are checked from left to
     * right, starting at the first column.
     * <p/>
     * The cell values can be specified as regular expressions.
     *
     * @param tableName          Name of the table.
     * @param row                The row number to be asserted.
     * @param expectedCellValues The expected cell values.
     */
    public void assertTableRow(String tableName, int row, String[] expectedCellValues);

    /**
     * Checks whether the text of a specified text component matches the specified value.
     * The expected text can be specified as a regular expression.
     *
     * @param textComponentName Name of the text component.
     * @param expectedText      The expected text.
     */
    public void assertText(String textComponentName, String expectedText);

    /**
     * Checks whether the specified OGNL expression is true.
     *
     * @param componentName  Name of the component.
     * @param ognlExpression The ognl expression used to assert.
     */
    public void assertTrue(String componentName, String ognlExpression);

    /**
     * Checks whether the specified OGNL expression is false.
     *
     * @param componentName  Name of the component.
     * @param ognlExpression The ognl expression used to assert.
     */
    public void assertFalse(String componentName, String ognlExpression);

    /**
     * Check whether a component with the specified name is enabled.
     *
     * @param componentName Name of the component.
     */
    public void assertEnabled(String componentName);

    /**
     * Check whether a component with the specified name is enabled.
     *
     * @param componentName Name of the component.
     */
    public void assertDisabled(String componentName);

    /**
     * * Check whether a toggle button with the specified name is selected.
     *
     * @param toggleButtonName Name of the toggle button.
     */
    public void assertToggleButtonSelected(String toggleButtonName);

    /**
     * Check whether a toggle button with the specified name is not selected.
     *
     * @param toggleButtonName Name of the toggle button.
     */
    public void assertToggleButtonNotSelected(String toggleButtonName);

    /**
     * Check whether a check box with the specified name is selected.
     *
     * @param checkBoxName Name of the check box.
     */
    public void assertCheckBoxSelected(String checkBoxName);

    /**
     * Check whether a check box with the specified name is not selected.
     *
     * @param checkBoxName Name of the check box.
     */
    public void assertCheckBoxNotSelected(String checkBoxName);

    /**
     * Check whether a radio button with the specified name is selected.
     *
     * @param radioButtonName Name of the radio button.
     */
    public void assertRadioButtonSelected(String radioButtonName);

    /**
     * Check whether a radio button with the specified name is not selected.
     *
     * @param radioButtonName Name of the radio button.
     */
    public void assertRadioButtonNotSelected(String radioButtonName);

    /**
     * Checks whether the text of a specified label component matches the specified value.
     *
     * @param expectedText The expected value in the label.
     */
    public void assertLabel(String expectedText);

    /**
     * Cancels a table edit.
     *
     * @param tableName Name of the table on which to cancel edit.
     */
    public void cancelTableEdit(String tableName);

    /**
     * Click on a specified button.
     *
     * @param buttonName Name of the button to be clicked.
     */
    public void clickButton(String buttonName);

    /**
     * Click on a specified checkbox.
     * The check box will be selected or deselected based on the isChecked value.
     *
     * @param checkBoxName The name of the check box.
     * @param isChecked    true if check box is expected to be checked, false otherwise.
     */
    public void clickCheckBox(String checkBoxName, boolean isChecked);

    /**
     * Click on the specified radio button.
     *
     * @param radioButtonName The name of the radio button to be clicked.
     */
    public void clickRadioButton(String radioButtonName);

    /**
     * Click on a specified column of a table header
     *
     * @param tableHeaderName The name of the table header.
     * @param tableColumnName The name of the table column name.
     */
    public void clickTableHeader(String tableHeaderName, String tableColumnName);

    /**
     * Close an internal frame with the specified title.
     * The title can be specified as a regular expression.
     *
     * @param internalFrameTitle The title of the internal frame to be closed.
     */
    public void closeInternalFrame(String internalFrameTitle);

    /**
     * Close all open dialogs recursively until no more dialogs are open.
     * This method can be used to remove unexpected dialogs at the end of a test.
     */
    public void closeAllDialogs();

    /**
     * Wait for a specified duration (in milliseconds)
     *
     * @param milliseconds The duration to wait in milliseconds.
     */
    public void delay(int milliseconds);

    /**
     * Wait for a dialog with a specified title to be closed.
     * <p/>
     * This method waits for 10 seconds in case the dialog is found to be open.
     * (The wait time will be made configurable in a future release).
     * The title can be specified as a regular expression.
     *
     * @param dialogTitle The title of the dialog to be waited for.
     */
    public void dialogClosed(String dialogTitle);

    /**
     * Wait for a dialog with a specified title to be opened.
     * <p/>
     * This function waits for 10 seconds in case the dialog is not found to be open.
     * (The wait will be made configurable in a future release).
     * The title can be specified as a regular expression.
     *
     * @param dialogTitle The title of the dialog to be waited for.
     */
    public void dialogShown(String dialogTitle);

    /**
     * Double click on a table row of a specified table.
     *
     * @param tableName The name of the table.
     * @param rowIndex  The row of the table.
     */
    public void doubleClickTableRow(String tableName, int rowIndex);

    /**
     * Double click on a list item of a specified list.
     *
     * @param listName  The name of the list.
     * @param itemIndex The row of the list.
     */
    public void doubleClickList(String listName, int itemIndex);

    /**
     * Double clicks on a tree. Supports regular expressions.
     * <p/>
     * Example: doubleClickTree("tree_name", new String[]{"top level", "/.*level/", "third level"});
     * </p>
     *
     * @param treeName     The name of the tree.
     * @param pathElements The array of the path elements of the tree.
     */
    public void doubleClickTree(String treeName, String[] pathElements);

    /**
     * Enter the specified text into a specified text field.
     *
     * @param textFieldname The name of the text field.
     * @param text          The text to be entered into the text field.
     */
    public void enterText(String textFieldname, String text);

    /**
     * Edit the table cell, identified by the row and the column numbers provided.
     *
     * @param tableName The name of the table.
     * @param row       The row number of the cell.
     * @param column    The column number of the cell.
     */
    public void editTableCell(String tableName, int row, int column);

    /**
     * Checks that an internal frame with the specified title has been shown
     * The title can be specified as a regular expression.
     *
     * @param internalFrameTitle The title of the internal frame.
     */
    public void internalFrameShown(String internalFrameTitle);

    /**
     * Enter a keystroke with the specified modifiers.
     * <p/>
     * Example: keyStroke("Ctrl+Alt 0") for the modifiers "Control" & "Alt" and the key "0".
     * </p>
     *
     * @param keyModifierAndKeyCodeText The key modifier and the key code given as a string separated by a space.
     */
    public void keyStroke(String keyModifierAndKeyCodeText);

    /**
     * Navigate to a specified path in a menu item.
     * Both menu bars and popup menus are supported.
     * <p/>
     * The path needs to be specified as a string delimited by the ">" character.
     * <p/>
     * Example: "first level>second level>item".
     * </p>
     * <p/>
     * This function does not have regular expression support.
     *
     * @param pathString The delimited path string.
     */
    public void navigate(String pathString);

    /**
     * Right click on a tree item of the specified tree. Supports regular expressions
     * <p/>
     * Example: rightClickTree("tree_name", new String[]{"top level", "/.*level/", "third level"});
     * </p>
     *
     * @param treeName     The name of the tree.
     * @param pathElements The array of the path elements of the tree.
     */
    public void rightClickTree(String treeName, String[] pathElements);

    /**
     * Right click on a list item of a specified list.
     *
     * @param listName  The name of the list.
     * @param itemIndex The index of the item in the list.
     */
    public void rightClickList(String listName, int itemIndex);

    /**
     * Right click on a table row of a specified table.
     *
     * @param tableName The name of the table.
     * @param rowIndex  The index of the table row to right clicked.
     */
    public void rightClickTableRow(String tableName, int rowIndex);

    /**
     * Select a specified value from a specified combo box.
     *
     * @param comboBoxName The name of the combo box.
     * @param value        The value to be selected in the combo box.
     */
    public void selectDropDown(String comboBoxName, String value);

    /**
     * Select a specified file in a file chooser.
     *
     * @param filePath The path of the file to be choosen.
     */
    public void selectFile(String filePath);

    /**
     * Select multiple files in a file chooser.
     *
     * @param filePaths The array of the file paths.
     */
    public void selectFiles(String[] filePaths);

    /**
     * Select a specified value in a specified list.
     *
     * @param listName     The name of the list.
     * @param listElements The array of the list elements to be selected.
     */
    public void selectList(String listName, String[] listElements);

    /**
     * Selects the specified rows in a table.
     * <p/>
     * Example: selectTableRow("table_name" , new int[]{1,2,3});
     * </p>
     *
     * @param tableName The name of the table.
     * @param rows      The arary of rows to be selected.
     */
    public void selectTableRow(String tableName, int[] rows);

    /**
     * Select a specified tree path. Supports regular expressions
     * <p/>
     * Example: selectTree("tree_name", new String[] {"top level", "/.*level/", "third level"});
     * </p>
     *
     * @param treeName     The name of the tree.
     * @param pathElements The array of the path elements.
     */
    public void selectTree(String treeName, String[] pathElements);

    /**
     * Stop editing a specified table.
     *
     * @param tableName The name of the table.
     */
    public void stopTableEdit(String tableName);

    /**
     * Switch to a tab with the specified title of a specified tabbed pane.
     * The title can be specified as a regular expression.
     *
     * @param tabPaneTitle The title of the tabbed pane.
     * @param tabTitle     The title of the tab.
     */
    public void switchTab(String tabPaneTitle, String tabTitle);

    /**
     * Moves a specified slider to the specified position
     *
     * @param sliderName The name of the slider.
     */
    public void moveSlider(String sliderName, int position);

    /**
     * Finishes the current test session.
     */
    public void finishTest();
}
